### PR TITLE
Created TestController, added/updated necessary methods in TestService

### PR DESCRIPTION
**Pull request by luinnarn**

### CODE REVIEW

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+@RestController
+public class TestController {
+
+    @Autowired
+    private TestService testService;
+
+    @RequestMapping(value = "/createtest/{formulaId}/{userId}",
+                    method = RequestMethod.POST)
+    public ResponseEntity<TestDto> createTest(@PathVariable("formulaId") Long formulaId,
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please make the path `/test/create` and pass `formulaId` and `userId` as part of the request body.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    public ResponseEntity<TestDto> getTestById(@PathVariable("testId") Long testId) {
+        Test test = testService.findOne(testId);
+        TestDto dto = new TestDto(test);
+        return new ResponseEntity<TestDto>(dto, HttpStatus.OK);
+    }
+
+    @RequestMapping(value = "/tests",
+                    method = RequestMethod.GET)
+    public ResponseEntity<List<TestDto>> getTests() {
+        List<Test> testList = testService.findAll();
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please rename to `tests` - it is not a good practice to add type info to the variable name.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        List<TestDto> testDtoList = new ArrayList<TestDto>();
+        for (Test test : testList) {
+            testDtoList.add(new TestDto(test));
+        }
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: We are using Java 8 in the project. Feel free to get to know it a bit, if you are unfamiliar with its features. :)

The upper 4 lines can be written as:
```
List<TestDto> testDtos = tests.stream().map(TestDto::new).collect(Collectors.toList());
```

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        return new ResponseEntity<List<TestDto>>(testDtoList, HttpStatus.OK);
+    }
+
+    @RequestMapping(value = "/users/{userId}/tests",
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please change mapping to `/tests/{userId}`

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                    method = RequestMethod.GET)
+    public ResponseEntity<List<TestDto>> getTestsByUserId(@PathVariable("userId") Long userId) {
+        List<Test> testList = testService.findTestsByUserId(userId);
+        List<TestDto> testDtoList = new ArrayList<TestDto>();
+        for (Test test : testList) {
+            testDtoList.add(new TestDto(test));
+        }
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: See my comment above about replacing the upper 4 lines.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        return new ResponseEntity<List<TestDto>>(testDtoList, HttpStatus.OK);
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please use diamond operator instead of explicitly stating List type, e.g.,
```
return new ResponseEntity<List<>>(testDtoList, HttpStatus.OK);
```

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java [hunk 7]
```diff
@@ -75,6 +84,10 @@ public Test findOne(Long id) {
     return testRepository.findAll();
   }
 
+  @Override
+  @Transactional(readOnly = true)
+  public List<Test> findTestsByUserId(Long userId) { return userRepository.getOne(userId).getTest(); }
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: Please do not inline method body with method header.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        TestDto dto = new TestDto(test);
+        return new ResponseEntity<TestDto>(dto, HttpStatus.OK);
+    }
+
+    @RequestMapping(value = "/tests/{testId}",
+                    method = RequestMethod.GET)
+    public ResponseEntity<TestDto> getTestById(@PathVariable("testId") Long testId) {
+        Test test = testService.findOne(testId);
+        TestDto dto = new TestDto(test);
+        return new ResponseEntity<TestDto>(dto, HttpStatus.OK);
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please use diamond operator instead.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                    method = RequestMethod.GET)
+    public ResponseEntity<List<TestDto>> getTestsByUserId(@PathVariable("userId") Long userId) {
+        List<Test> testList = testService.findTestsByUserId(userId);
+        List<TestDto> testDtoList = new ArrayList<TestDto>();
+        for (Test test : testList) {
+            testDtoList.add(new TestDto(test));
+        }
+        return new ResponseEntity<List<TestDto>>(testDtoList, HttpStatus.OK);
+    }
+}
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please format the file.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import eu.execom.testgenerator.core.model.Test;
+import eu.execom.testgenerator.core.model.User;
+import eu.execom.testgenerator.core.service.FormulaService;
+import eu.execom.testgenerator.core.service.UserService;
+import eu.execom.testgenerator.web.dto.TestDto;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.method.P;
+import org.springframework.web.bind.annotation.*;
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please do not use group imports. Setup your IDE to group imports only if there are more than 50 from the same package.

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestDto.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+@NoArgsConstructor
+@Data
+public class TestDto {
+    private String applicantName;
+    private String testQuestions;
+
+    public TestDto(Test test) {
+        this.applicantName = test.getApplicant().getName();
+        this.testQuestions = test.getTestQuestions().toString();
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestDto.java]: Please make this similar to the `Test` class, just without the information about the database mapping and use id's instead of whole classes references.

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java [hunk 12]
```diff
@@ -16,9 +20,14 @@
     private Long reviewerId;
     private Long formulaId;
 
-
     public TestResponse(Test test) {
         this.applicantName = test.getApplicant().getName();
-        this.testQuestions = test.getTestQuestions().toString();
+        ObjectMapper mapper = new ObjectMapper();
+        try {
+            this.testQuestions = mapper.writeValueAsString(
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java]: Please do not serialize manually, Jackson does that for you automatically when you send back the response.

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestQuestionDto.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
+package eu.execom.testgenerator.web.dto;
+
+import eu.execom.testgenerator.core.model.TestQuestion;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@Data
+public class TestQuestionDto {
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestQuestionDto.java]: TestQuestion is a structure introduced to represent a join table between Test and a Question. When you send the information to the UI (or wherever) through controller, you just need the information from Question, so, you should write your TestResponse class similar to this:
```
public class TestResponse {
  // everything else
  private List<QuestionDto> questions;
}

public class QuestionDto {
  
  private Long id;
  private QuestionType type;
  private String text;
  // investigate what is the best way to send an image to the UI
  private byte[] image;
}

```

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+@RestController
+public class TestController {
+
+    @Autowired
+    private TestService testService;
+
+    @RequestMapping(value = "/tests/create",
+                    method = RequestMethod.POST,
+                    headers = "Content-Type=application/json;charset=utf-8")
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please rather use value for `consumes` and `produces`.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+    @RequestMapping(value = "/tests/{userId}",
+                    method = RequestMethod.GET)
+    public ResponseEntity<List<TestResponse>> getTestsByUserId(@PathVariable("userId") Long userId) {
+        List<Test> tests = testService.findTestsByUserId(userId);
+        List<TestResponse> testResponses = tests.stream().map(TestResponse::new).collect(Collectors.toList());
+        return new ResponseEntity<>(testResponses, HttpStatus.OK);
+    }
+
+    @RequestMapping(value = "/deletetest/{testId}",
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please map this method to `/test/{testId}` and `method = RequestMethod.DELETE`.

**Comment by stanislavgrujic** [comment]
Please resolve merge conflicts.

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java (hunk 7 above)
Comment (stanislavgrujic) [reply core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: This comment is still not resolved.

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java [hunk 16]
```diff
@@ -22,10 +24,17 @@
...
   private TestRepository testRepository;
+  @Autowired
+  private UserRepository userRepository;
+  @Autowired
+  private FormulaRepository formulaRepository;
 
   @Override
   @Transactional
-  public Test create(Formula formula, User user) {
+  public Test create(Long formulaId, Long userId) {
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: Why did you change the method signature? Having specific types in the method header instead of Long values give s you more type safety (e.g., with this method signature, you can easily switch `formulaId` and `userId` when calling the method, while it cannot happen when you use specific types).

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    private Long id;
+    private boolean deleted;
+    private Long applicantId;
+    private Long reviewerId;
+    private Long formulaId;
+
+    public TestResponse(Test test) {
+        this.applicantName = test.getApplicant().getName();
+        this.questions = test.getTestQuestions().stream().map(TestQuestion::getQuestion).map(QuestionDto::new).collect(
+                Collectors.toList());
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java]: How do you populate the other fields in this class with this constructor?

Comment (luinnarn) [reply web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java]: Is it better for the TestResponse object to hold a reference to the Test object as the constructor arg (private Test test ... this.test == test), or to populate all the fields manually in the constructor?

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java [hunk 18]
```diff
@@ -22,81 +22,83 @@
...
-  @Autowired
-  private UserRepository userRepository;
-  @Autowired
-  private FormulaRepository formulaRepository;
+    @Autowired
+    private TestRepository testRepository;
+    @Autowired
+    private UserRepository userRepository;
+    @Autowired
+    private FormulaRepository formulaRepository;
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: The spacing should be 2 spaces for tab character. Is something changed in your formatter?

Comment (luinnarn) [reply core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: I suspect my formatter is misbehaving... I will check it out.

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java [hunk 19]
```diff
@@ -84,8 +86,24 @@ public Test findOne(Long id) {
 
   @Override
   @Transactional(readOnly = true)
-  public List<Test> findTestsByUserId(Long userId) {
-    return userRepository.getOne(userId).getTest();
+  public List<Test> findTestsByApplicantId(Long userId) {
+    User user = userRepository.getOne(userId);
+    if (user != null && user.getRole() == UserRole.APPLICANT) {
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: Shouldn't call to `getOne()` already throw an `EntityNotFoundException`, if an entity is not found, is null check really needed?

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java [hunk 20]
```diff
@@ -84,8 +86,24 @@ public Test findOne(Long id) {
...
+      return userRepository.getOne(userId).getApplicantTests();
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: Do you have to make another call to the `userRepository` to fetch user one more time, just in order to get his tests? Can't you just `return user.getApplicantTests()`?

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 21]
```diff
@@ -30,20 +34,29 @@
...
+    if (userService.findByEmail(testRequest.getUserEmail()) != null) {
+      user = userService.findByEmail(testRequest.getUserEmail());
+    } else {
+      user = new User();
+      user.setRole(UserRole.APPLICANT);
+      user.setName(testRequest.getUserName());
+      user.setEmail(testRequest.getUserEmail());
+      userService.save(user);
+    }
+    Test test = testService.create(formulaService.findOne(testRequest.getFormulaId()), user);
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: This logic should all go into service, not in the controller.

Comment (stanislavgrujic) [reply web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Think of a controller just as a proxy to the service.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 22]
```diff
@@ -54,29 +67,32 @@
...
+    List<Test> tests = new ArrayList<>();
     if (userId == 0) {
       tests = testService.findAll();
-    } else {
-      tests = testService.findTestsByUserId((long) userId);
+    } else if (userService.findOne((long) userId).getRole() == UserRole.APPLICANT) {
+      tests = testService.findTestsByApplicantId((long) userId);
+    } else if (userService.findOne((long) userId).getRole() == UserRole.REVIEWER) {
+      tests = testService.findTestsByReviewerId((long) userId);
     }
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please move logic to the service

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java (hunk 9 above)
Comment (stanislavgrujic) [reply web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Once you move the logic to the service, please add unit tests for it.

File: core/src/main/java/eu/execom/testgenerator/core/model/User.java [hunk 23]
```diff
@@ -51,6 +51,23 @@
...
+  @OneToMany(mappedBy = "applicant",
+             cascade = CascadeType.ALL)
+  private List<Test> applicantTests;
+
+  @OneToMany(mappedBy = "reviewer",
+             cascade = CascadeType.ALL)
+  private List<Test> reviewerTests;
+
+  public boolean isApplicant() {
+    return (this.getRole() == UserRole.APPLICANT ? true : false);
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/model/User.java]: Please rewrite the method body as (and the ones below):
```
return this.role == UserRole.APPLICANT;
```

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java [hunk 24]
```diff
@@ -75,6 +84,26 @@ public Test findOne(Long id) {
...
+      User user = userRepository.getOne(userId);
+      if (user.isReviewer()) {
+        return user.getReviewerTests();
+      } else if (user.isApplicant()) {
+        return user.getApplicantTests();
+      } else {
+        return findAll();
+      }
+    } else {
+      throw new EntityNotFoundException();
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: This branch is probably never reachable since `userRepository.getOne()` will throw the exception if the user is not found.

File: core/src/main/java/eu/execom/testgenerator/core/service/UserServiceImpl.java [hunk 25]
```diff
@@ -20,25 +21,47 @@ public User save(User user) {
...
     userRepository.delete(id);
   }
+
+  @Override
+  @Transactional
+  public User getUserByNameAndEmail(String name, String email) {
+    User user;
+    if (findByEmail(email) != null) {
+      user = findByEmail(email);
+    } else {
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/UserServiceImpl.java]: This is very inefficient - you will call `findByEmail()` method two times.
Please rewrite as
```
User user = findByEmail(email);
if (user == null) {
  // the rest of logic
}
```

File: core/src/main/java/eu/execom/testgenerator/core/service/UserServiceImpl.java [hunk 26]
```diff
@@ -20,25 +21,47 @@ public User save(User user) {
...
+      user = new User();
+      user.setRole(UserRole.APPLICANT);
+      user.setName(name);
+      user.setEmail(email);
+      save(user);
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/UserServiceImpl.java]: Are we sure that we want the method getByEmail to actually create a new user?

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+  @Autowired
+  private UserService userService;
+
+  @RequestMapping(value = "/create",
+                  method = RequestMethod.POST,
+                  consumes = MediaType.APPLICATION_JSON_VALUE,
+                  produces = MediaType.APPLICATION_JSON_VALUE)
+  public ResponseEntity<TestResponse> createTest(
+      @RequestBody
+          TestRequest testRequest) {
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please inline these parameters.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+    TestResponse response = new TestResponse(test);
+    return new ResponseEntity<>(response, HttpStatus.OK);
+  }
+
+  @RequestMapping(value = "/{testId}",
+                  method = RequestMethod.GET,
+                  produces = MediaType.APPLICATION_JSON_VALUE)
+  public ResponseEntity<TestResponse> getTestById(
+      @PathVariable("testId")
+          Long testId) {
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please inline these parameters with the method header.

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java (hunk 24 above)
Comment (luinnarn) [reply core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: This branch should cover cases where a nonexistent id is passed as `RequestParam`, in order to avoid returning all tests in such cases.

File: core/src/main/java/eu/execom/testgenerator/core/service/UserServiceImpl.java (hunk 26 above)
Comment (luinnarn) [reply core/src/main/java/eu/execom/testgenerator/core/service/UserServiceImpl.java]: The method `getUserByNameAndEmail` is being called during test creation, at which moment a new user should be created if not found in the database, if I understood the wiki correctly... What's missing at this point is functionality to check if there is an existing user with the same full name, in which case the system asks if the user has already taken the test, and asks for the e-mail which was used previously

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java (hunk 27 above)
Comment (luinnarn) [reply web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: That darn formatter is at it again

File: core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java (hunk 24 above)
Comment (luinnarn) [reply core/src/main/java/eu/execom/testgenerator/core/service/TestServiceImpl.java]: ... if I'm not mistaken. I'll try to find a better solution.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+    TestResponse response = new TestResponse(test);
+    return new ResponseEntity<>(response, HttpStatus.CREATED);
+  }
+
+  @RequestMapping(method = RequestMethod.GET,
+                  produces = MediaType.APPLICATION_JSON_VALUE)
+  public ResponseEntity<List<TestResponse>> getTests(@RequestParam(name = "userId",
+                                                                   defaultValue = "0") int userId) {
+    List<Test> tests;
+    tests = testService.findTestsByUserId((long) userId);
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please do not split the variable declaration and definition unless there is really need to do that -> make this two lines a single one.

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@NoArgsConstructor
+@Data
+public class TestResponse {
+
+  private Long id;
+  private boolean deleted;
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java]: This field should not exist in the response. If a test is deleted, it should not be retrieved from the database and the UI should not know about it.

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+  private Long formulaId;
+
+  private List<QuestionDto> questions;
+
+  public TestResponse(Test test) {
+    if (test != null) {
+      this.id = test.getId();
+      this.deleted = test.isDeleted();
+      this.applicantId = test.getApplicant().getId();
+      this.reviewerId = ((test.getReviewer() == null)? null : test.getReviewer().getId());
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestResponse.java]: Please fix formatting.

File: web/src/test/java/eu/execom/testgenerator/web/controller/TestControllerTest.java [hunk 32]
```diff
@@ -0,0 +1,139 @@
...
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@RunWith(MockitoJUnitRunner.class)
+@TestExecutionListeners({})
```
Comment (stanislavgrujic) [inline web/src/test/java/eu/execom/testgenerator/web/controller/TestControllerTest.java]: Please remove this annotation, it's not used.

File: web/src/test/java/eu/execom/testgenerator/web/controller/TestControllerTest.java [hunk 33]
```diff
@@ -0,0 +1,139 @@
...
+    mockMvc = MockMvcBuilders.standaloneSetup(testController).addFilters(new CorsFilter()).build();
+  }
+
+  @org.junit.Test
+  public void shouldGetAllTests() throws Exception {
+    User applicant = applicant(2L);
+    Formula formula = formula("FormulaName", 3L);
+    Test test1 = test(1L, formula, applicant, new ArrayList<TestQuestion>());
+    Test test2 = test(2L, formula, applicant, new ArrayList<TestQuestion>());
+    Test test3 = test(3L, formula, applicant, new ArrayList<TestQuestion>());
```
Comment (stanislavgrujic) [inline web/src/test/java/eu/execom/testgenerator/web/controller/TestControllerTest.java]: If you're always passing an empty ArrayList you can also internalize it in the method body.

File: web/src/test/java/eu/execom/testgenerator/web/controller/TestControllerTest.java [hunk 34]
```diff
@@ -0,0 +1,139 @@
...
+    User user = new User();
+    user.setId(3L);
+
+    //when
+    when(formulaService.findOne(1L)).thenReturn(formula);
+    when(userService.getUserByNameAndEmail(anyString(), anyString())).thenReturn(user);
+
+    //then
+    mockMvc.perform(post(URL_PREFIX + "/create").content(
+        "{\"formulaId\" : 1,\"userName\" : \"Slavko Stimac\",\"userEmail\" : \"salas@umalomritu\"}")
```
Comment (stanislavgrujic) [inline web/src/test/java/eu/execom/testgenerator/web/controller/TestControllerTest.java]: If we add e-mail validatiom, this request will fail it. Please use proper email addresses.

File: web/src/main/java/eu/execom/testgenerator/web/dto/TestRequest.java [hunk 35]
```diff
@@ -0,0 +1,12 @@
...
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@Data
+public class TestRequest {
+  private Long formulaId;
+  private String userName;
+  private String userEmail;
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/TestRequest.java]: Please add request validation.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 36]
```diff
@@ -0,0 +1,71 @@
...
+  @Autowired
+  private FormulaService formulaService;
+  @Autowired
+  private UserService userService;
+
+  @RequestMapping(value = "/create",
+                  method = RequestMethod.POST,
+                  consumes = MediaType.APPLICATION_JSON_VALUE,
+                  produces = MediaType.APPLICATION_JSON_VALUE)
+  public ResponseEntity<TestResponse> createTest(@RequestBody TestRequest testRequest) {
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please add @Valid annotation for the request body, and also add validation annotations to the `TestRequest`.

File: core/src/test/java/eu/execom/testgenerator/core/service/TestServiceTest.java [hunk 37]
```diff
@@ -54,10 +73,23 @@ public void setUp() {
...
+      @Override
+      public List<Test> answer(InvocationOnMock invocationOnMock) throws Throwable {
+        return allTests;
+      }
+    });
+    given(userRepository.getOne(any(Long.TYPE))).willAnswer(new Answer<User>() {
+      @Override
+      public User answer(InvocationOnMock invocationOnMock) throws Throwable {
+        return user;
+      }
```
Comment (stanislavgrujic) [inline core/src/test/java/eu/execom/testgenerator/core/service/TestServiceTest.java]: Please use lambda expressions and Java8 features to define the mocks behavior.

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 38]
```diff
@@ -0,0 +1,75 @@
...
+public class TestController {
+
+  @Autowired
+  private TestService testService;
+  @Autowired
+  private FormulaService formulaService;
+  @Autowired
+  private UserService userService;
+
+  private final Logger logger = LoggerFactory.getLogger(TestController.class);
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Please use Lombok annotation `@Slf4j`.

File: web/src/test/java/eu/execom/testgenerator/web/controller/TestControllerTest.java [hunk 39]
```diff
@@ -0,0 +1,133 @@
...
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TestControllerTest {
+
+  private static final String URL_PREFIX = "/tests";
+
+  private static final MediaType JSON_CONTENT_TYPE = new MediaType(MediaType.APPLICATION_JSON.getType(),
+      MediaType.APPLICATION_JSON.getSubtype(), Charset.forName("utf8"));
```
Comment (stanislavgrujic) [inline web/src/test/java/eu/execom/testgenerator/web/controller/TestControllerTest.java]: Can't you just use `MediaType.APPLICATION_JSON` in code, instead of having to create this constant?
